### PR TITLE
GitHub Actions & Dev Tools Cleanup + Python 3.11 Support

### LINKED ISSUE #1007: [Bug] Parsons 3.1.0 targets yanked 4.0.0 version of braintree with critical bugs
Parsons 3.1.0 depends on braintree 4.0.0 which has been yanked due to critical bugs.

## Detailed Description

When I install parsons 3.1.0 via pip, I get this message:
> WARNING: The candidate selected for download or install is a yanked version: 'braintree' candidate (version 4.0.0 at https://files.pythonhosted.org/packages/a5/1e/641306eec960161028e0a30d2f635abdfbdb89029330df1701bc106e5d8c/braintree-4.0.0-py2.py3-none-any.whl (from https://pypi.org/simple/braintree/))
Reason for being yanked: critical bugs

When I install it via [uv](https://github.com/astral-sh/uv) it doesn't even install for the same reason.

## To Reproduce

Create a new virtual environment and run `pip install parsons==3.1.0`

## Your Environment

* Version of Parsons used (if you know): **v3.1.0**
* Environment name and version (e.g. Chrome 39, node.js 5.4): 
* Operating System and version (desktop or mobile): **Pop!_OS 22.04 LTS**

## Additional Context

In a fork I'm playing around with, braintree==4.26.0 and google-cloud-bigquery==3.18.0 were needed to get pytests passing. I'm not sure why that isn't the case on this repo but it seems like these versions might be safe to update.

## Priority

For me it's low because I don't use braintree but "critical bugs" makes it sound rather serious for people who use that integration.

### PR DESCRIPTION
**Pull request by bmos**
I'm not really sure how these changes could be merged nicely with other open PRs.
I'm happy to restructure it however is desired, so perhaps treat this as a sort of menu.

## What's Included

- Simplified the linting workflow to use a single file for both mac and linux testing.
- Use the rust-based package installer [uv](https://github.com/astral-sh/uv) to significantly reduce the GitHub actions runtime to allow a second python version in macOS testing.
- Functionalized `__init__.py` (which is kind of pointless but I did it while debugging an error I was getting) and fixed a typo in it (one of the tuples was a list instead).
- Fix import of HttpError in gmail.py (which for some reason your pytest workflow was not catching).
- Updated packages that were needed to get pytest to pass or were missing patch-level updates.
- Restructured limited dependencies mode to use the same dependency versions as the usual install.
- Resolve the TypeError raised when running pytest under Python 3.11. Required [upstream fix](https://github.com/petl-developers/petl/pull/656).
- Flake8 was set to maintain a 100 character max line length, but Black didn't have a line length option configured which meant it was defaulting to a much lower 88. These are now both 100.
- Add flake8, pytest, and black configurations to pyproject.toml rather than having a mix of separate config files and command-line args -- also standardized the commands used in all places.
- Updated Black and Flake8 to latest versions and applied their updated linting suggestions.
- Pytest now finishes with 3 warnings instead of 13.

## Notes / Considerations

- Removed support for python 3.7 as it is not supported with uv. This could be added back as a supported version along with its own test workflow or no test workflow, if preferred. I know dropping 3.7 was recently discussed in #932.
- Closes #1007 by using a version of braintree that has not been yanked

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,49 +1,48 @@
...
+braintree==4.26.0
+bs4==0.0.1
+censusgeocode==0.4.3.post1
+civis==1.16.1
+dbt_redshift==1.4.0
+docutils<0.18,>=0.14
 facebook-business==13.0.0
 google-api-python-client==1.7.7
+google-auth==2.6.2
+google-cloud-bigquery==3.18.0
```
Comment (bmos) [inline requirements.txt:14]: updated from 3.4.0

File: requirements.txt [hunk 2]
```diff
@@ -1,49 +1,48 @@
...
-slackclient==1.3.0
-psycopg2-binary==2.9.3
-xmltodict==0.11.0
-gspread==3.7.0
-oauth2client==4.1.3
-google-auth==2.6.2
+braintree==4.26.0
+bs4==0.0.1
+censusgeocode==0.4.3.post1
+civis==1.16.1
```
Comment (bmos) [inline requirements.txt:8]: updated from 1.16.0

File: requirements.txt [hunk 3]
```diff
@@ -1,49 +1,48 @@
...
+google-cloud-storage-transfer==1.9.1
+google-cloud-storage==2.10.0
 google-resumable-media==2.3.2
 grpcio==1.53.2
-httplib2==0.19.0
-validate-email==1.3
-paramiko==3.4.0
-xmltodict==0.11.0
+gspread==3.7.0
+httplib2==0.22.0
```
Comment (bmos) [inline requirements.txt:20]: updated from 0.19.0

File: requirements.txt [hunk 4]
```diff
@@ -1,49 +1,48 @@
...
-suds-py3==1.3.4.0
-newmode==0.1.6
 mysql-connector-python==8.0.18
-braintree==4.0.0
-python-dateutil==2.8.2
-azure-storage-blob==12.3.2
+newmode==0.1.6
+oauth2client==4.1.3
+paramiko==3.4.0
+petl==1.7.14
```
Comment (bmos) [inline requirements.txt:26]: updated from 1.6.8

File: requirements.txt (hunk 3 above)
Comment (bmos) [inline requirements.txt:20]: updated httplib2 from 0.19.0

**Comment by austinweisgrau** [comment]
I think this should be split into several different PRs for the different components here. Each PR should explain the motivation for the relevant changes and what is affected.

I see these are separate PR scopes:
- linting changes
- CICD changes
- package management changes
- 3.11 compatibility
- bug fixes

**Comment by austinweisgrau** [comment]
Good work though, these look mostly like nice fixes. I think we'd want to discuss pros and cons of switching to uv, and I think it would be preferable to stick to the black line length default of 88 rather than switching it to the flake8 default of 100, but we should have those discussions in smaller scoped PRs.

**Comment by bmos** [comment]
Thanks for those notes, I will restructure this into the suggested categories.

> I think it would be preferable to stick to the black line length default of 88 rather than switching it to the flake8 default of 100

100 is not the default for flake8, that would be 79.
I went with 100 because the flake8 [config file](https://github.com/move-coop/parsons/blob/main/.flake8) in your repo says 100 and the CONTRIBUTING.md file also [includes it](https://github.com/move-coop/parsons/blob/main/CONTRIBUTING.md?plain=1#L95) in the flake8 command. If 88 is the desired value I can change those values of 100 to 88 -- standardization was my goal, not a longer line length.

**Comment by austinweisgrau** [comment]
> Thanks for those notes, I will restructure this into the suggested categories.
> 
> > I think it would be preferable to stick to the black line length default of 88 rather than switching it to the flake8 default of 100
> 
> 100 is not the default for flake8, that would be 79. I went with 100 because the flake8 [config file](https://github.com/move-coop/parsons/blob/main/.flake8) in your repo says 100 and the CONTRIBUTING.md file also [includes it](https://github.com/move-coop/parsons/blob/main/CONTRIBUTING.md?plain=1#L95) in the flake8 command. If 88 is the desired value I can change those values of 100 to 88 -- standardization was my goal, not a longer line length.

Oh word gotcha that makes sense! Yeah I think maintaining whatever the documented value is makes the most sense so that sounds right to me.

**Comment by bmos** [comment]
Does the major release branch need to be brought up to main?

**Comment by austinweisgrau** [comment]
It doth, yes. It will be eventually, PRs against it will look a little
messy until that happens.

**Pull request closed without merging**